aptos-move/framework/aptos-framework/sources/transaction_fee.move

```diff
@@ -1,15 +1,104 @@
+    const EINVALID_BURN_PERCENTAGE: u64 = 2;
+
+    /// Stores burn capability to burn the gas fees.
     struct AptosCoinCapabilities has key {
         burn_cap: BurnCapability<AptosCoin>,
     }
 
+    /// Stores information about the block proposer and the amount of fees
+    /// collected when executing the block.
+    struct CollectedFeesPerBlock has key {
+        amount: AggregatableCoin<AptosCoin>,
+        proposer: Option<address>,
+        burn_percentage: u8,
+    }
+
+    /// Initializes the resource storing information about gas fees collection and
+    /// distribution. Should be called by on-chain governance.
+    public fun initialize_fee_collection_and_distribution(aptos_framework: &signer, burn_percentage: u8) {
+        // Sanity checks.
+        system_addresses::assert_aptos_framework(aptos_framework);
+        assert!(
+            !exists<CollectedFeesPerBlock>(@aptos_framework),
+            error::already_exists(EALREADY_COLLECTING_FEES)
+        );
+        assert!(burn_percentage <= 100, error::out_of_range(EINVALID_BURN_PERCENTAGE));
+
+        // Make sure stakng module is aware of transaction fees collection.
+        stake::initialize_validator_fees(aptos_framework);
+
+        // Initially, no fees are collected and the block proposer is not set.
+        let zero = coin::initialize_aggregatable_coin(aptos_framework);
+        let info = CollectedFeesPerBlock {
+            amount: zero,
+            proposer: option::none(),
+            burn_percentage,
+        };
+        move_to(aptos_framework, info);
+    }
+
+    /// Registers the proposer of the block for gas fees collection. This function
+    /// can only be called at the beginning of the block.
+    public(friend) fun register_proposer_for_fee_collection(proposer_addr: address) acquires CollectedFeesPerBlock {
+        let collected_fees = borrow_global_mut<CollectedFeesPerBlock>(@aptos_framework);
+        let _ = option::swap_or_fill(&mut collected_fees.proposer, proposer_addr);
+    }
+
+    /// Burns a specified fraction of the coin.
+    fun burn_coin_fraction(coin: &mut Coin<AptosCoin>, burn_percentage: u8) acquires AptosCoinCapabilities {
+        assert!(burn_percentage <= 100, error::out_of_range(EINVALID_BURN_PERCENTAGE));
+
+        let collected_amount = coin::value(coin);
+        let amount_to_burn = (burn_percentage as u64) * collected_amount / 100;
+        let coin_to_burn = coin::extract(coin, amount_to_burn);
+        burn_collected_fee(coin_to_burn);
+    }
+
+    /// Calculates the fee which should be distributed to the block proposer at the
+    /// end of an epoch, and records it in the system. This function can only be called
+    /// at the beginning of the block or during reconfiguration.
+    public(friend) fun assign_or_burn_collected_fee() acquires AptosCoinCapabilities, CollectedFeesPerBlock {
+        let collected_fees = borrow_global_mut<CollectedFeesPerBlock>(@aptos_framework);
```

Comment:
Can we add a safeguard here if CollectedFeesPerBlock resource is not found? We can fallback to not doing anything. This shouldn't happen in practice (esp with multistep governance proposal support) but is possible in theory if the feature flag is enabled before the resource is initialized.

Reply:
Can you add a comment here to explain we wouldn't worry about this because the resource should be initialized along with enabling the feature flag?